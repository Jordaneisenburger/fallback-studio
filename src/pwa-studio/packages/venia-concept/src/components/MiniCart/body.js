import React, { useCallback } from 'react';
import { array, bool, func, object, shape, string } from 'prop-types';

import { mergeClasses } from '../../classify';
import LoadingIndicator from '../LoadingIndicator';

import defaultClasses from './body.css';
import EditItem from './editItem';
import EmptyMiniCartBody from './emptyMiniCartBody';
import ProductList from './productList';

const loadingIndicator = (
    <LoadingIndicator>{`Fetching Cart...`}</LoadingIndicator>
);

const Body = props => {
    // Props.
    const {
        beginEditItem,
        cartItems,
        closeDrawer,
        currencyCode,
        editItem,
        endEditItem,
        isCartEmpty,
        isEditingItem,
        isLoading,
        isUpdatingItem,
        removeItemFromCart,
        updateItemInCart
    } = props;

    // Members.
    const classes = mergeClasses(defaultClasses, props.classes);

    // Callbacks.
    const handleEditItem = useCallback(
        item => {
            beginEditItem(item);
        },
        [beginEditItem]
    );

    // Render.
    if (isLoading) {
        return loadingIndicator;
    }

    if (isCartEmpty) {
        return <EmptyMiniCartBody closeDrawer={closeDrawer} />;
    }

    if (isEditingItem) {
        return (
            <EditItem
                currencyCode={currencyCode}
                endEditItem={endEditItem}
                isUpdatingItem={isUpdatingItem}
                item={editItem}
                updateItemInCart={updateItemInCart}
            />
        );
    }

    return (
        <div className={classes.root}>
            <ProductList
                beginEditItem={handleEditItem}
                cartItems={cartItems}
                currencyCode={currencyCode}
                removeItemFromCart={removeItemFromCart}
            />
        </div>
    );
};

Body.propTypes = {
    beginEditItem: func.isRequired,
    cartItems: array,
    classes: shape({
        root: string
    }),
    closeDrawer: func,
    currencyCode: string,
    editItem: object,
    endEditItem: func,
    isCartEmpty: bool,
    isEditingItem: bool,
    isLoading: bool,
    isUpdatingItem: bool,
    removeItemFromCart: func,
    updateItemInCart: func
};

export default Body;
