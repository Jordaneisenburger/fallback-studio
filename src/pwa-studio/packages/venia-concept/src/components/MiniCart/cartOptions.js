import React, { Suspense, useCallback, useState } from 'react';
import { array, bool, func, number, shape, string } from 'prop-types';
import { Form } from 'informed';
import { Price } from '@magento/peregrine';

import { mergeClasses } from '../../classify';
import LoadingIndicator from '../LoadingIndicator';
import Button from '../Button';
import Quantity from '../ProductQuantity';
import appendOptionsToPayload from '../../util/appendOptionsToPayload';
import isProductConfigurable from '../../util/isProductConfigurable';

import defaultClasses from './cartOptions.css';

const Options = React.lazy(() => import('../ProductOptions'));

const loadingIndicator = (
    <LoadingIndicator>
        <span>{'Fetching Options...'}</span>
    </LoadingIndicator>
);

const isItemMissingOptions = (cartItem, configItem, numSelections) => {
    // Non-configurable products can't be missing options
    if (cartItem.product_type !== 'configurable') {
        return false;
    }

    // Configurable products are missing options if we have fewer
    // option selections than the product has options.
    const { configurable_options } = configItem;
    const numProductOptions = configurable_options.length;

    return numSelections < numProductOptions;
};

const CartOptions = props => {
    // Props.
    const {
        cartItem,
        configItem,
        currencyCode,
        endEditItem,
        isUpdatingItem,
        updateCart
    } = props;
    const { name, price, qty } = cartItem;

    // State.
    const [optionSelections, setOptionSelections] = useState(new Map());
    const [quantity, setQuantity] = useState(qty);

    // Callbacks.
    const handleSelectionChange = useCallback(
        (optionId, selection) => {
            setOptionSelections(
                new Map(optionSelections).set(
                    optionId,
                    Array.from(selection).pop()
                )
            );
        },
        [optionSelections]
    );
    const handleUpdateClick = useCallback(() => {
        const payload = {
            item: configItem,
            productType: configItem.__typename,
            quantity: quantity
        };

        if (isProductConfigurable(configItem)) {
            appendOptionsToPayload(payload, optionSelections);
        }

        updateCart(payload, cartItem.item_id);
    }, [cartItem, configItem, quantity, optionSelections, updateCart]);

    // Members.
    const classes = mergeClasses(defaultClasses, props.classes);
    const isMissingOptions = isItemMissingOptions(
        cartItem,
        configItem,
        optionSelections.size
    );
    const modalClass = isUpdatingItem ? classes.modal_active : classes.modal;

    // Render.
    const options = isProductConfigurable(configItem) ? (
        <Suspense fallback={loadingIndicator}>
            <section className={classes.options}>
                <Options
                    onSelectionChange={handleSelectionChange}
                    product={configItem}
                />
            </section>
        </Suspense>
    ) : null;

    return (
        <Form className={classes.root}>
            <div className={classes.focusItem}>
                <span className={classes.name}>{name}</span>
                <span className={classes.price}>
                    <Price currencyCode={currencyCode} value={price} />
                </span>
            </div>
            <div className={classes.form}>
                {options}
                <section className={classes.quantity}>
                    <h2 className={classes.quantityTitle}>
                        <span>Quantity</span>
                    </h2>
                    <Quantity initialValue={qty} onValueChange={setQuantity} />
                </section>
            </div>
            <div className={classes.save}>
                <Button onClick={endEditItem}>
                    <span>Cancel</span>
                </Button>
                <Button
                    priority="high"
                    onClick={handleUpdateClick}
                    disabled={isMissingOptions}
                >
                    <span>Update Cart</span>
                </Button>
            </div>
            <div className={modalClass}>
                <LoadingIndicator>Updating Cart</LoadingIndicator>
            </div>
        </Form>
    );
};

CartOptions.propTypes = {
    cartItem: shape({
        item_id: number.isRequired,
        name: string.isRequired,
        price: number.isRequired,
        qty: number.isRequired
    }),
    classes: shape({
        root: string,
        focusItem: string,
        price: string,
        form: string,
        quantity: string,
        quantityTitle: string,
        save: string,
        modal: string,
        modal_active: string,
        options: string
    }),
    configItem: shape({
        __typename: string,
        configurable_options: array
    }).isRequired,
    currencyCode: string,
    endEditItem: func.isRequired,
    isUpdatingItem: bool,
    updateCart: func.isRequired
};

export default CartOptions;
