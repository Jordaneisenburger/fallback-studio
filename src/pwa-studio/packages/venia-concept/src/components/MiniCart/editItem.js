import React, { useEffect } from 'react';
import { bool, func, object, string } from 'prop-types';

import { useQuery } from '@magento/peregrine';

import LoadingIndicator from '../LoadingIndicator';
import PRODUCT_DETAILS from '../../queries/getProductDetailByName.graphql';

import CartOptions from './cartOptions';

const loadingIndicator = (
    <LoadingIndicator>{`Fetching Item Options...`}</LoadingIndicator>
);

const EditItem = props => {
    // Props.
    const {
        currencyCode,
        endEditItem,
        isUpdatingItem,
        item,
        updateItemInCart
    } = props;

    // State / Hooks.
    const [queryResult, queryApi] = useQuery(PRODUCT_DETAILS);
    const { data, error, loading } = queryResult;
    const { runQuery, setLoading } = queryApi;

    // Members.
    const itemHasOptions = item && item.options && item.options.length > 0;

    // Run the query once on mount and again whenever the
    // item being edited changes.
    useEffect(() => {
        // We must have an item with options.
        if (!(item && itemHasOptions)) {
            return;
        }

        const fetchItemOptions = async () => {
            setLoading(true);

            await runQuery({
                variables: {
                    name: item.name,
                    onServer: false
                }
            });

            setLoading(false);
        };

        fetchItemOptions();
    }, [item, itemHasOptions, runQuery, setLoading]);

    /*
     * Render.
     */

    // We must have an item to edit.
    if (!item) {
        // TODO: log an error? Pop a toast?
        return null;
    }

    // This is a non-configurable item, just set configItem to an empty object.
    if (!itemHasOptions) {
        return (
            <CartOptions
                cartItem={item}
                configItem={{}}
                currencyCode={currencyCode}
                endEditItem={endEditItem}
                isUpdatingItem={isUpdatingItem}
                updateCart={updateItemInCart}
            />
        );
    }

    if (error) {
        return <span>Unable to fetch item options.</span>;
    }
    if (loading || !data) {
        return loadingIndicator;
    }

    // We do have this item's data.
    const itemWithOptions = data.products.items[0];

    return (
        <CartOptions
            cartItem={item}
            configItem={itemWithOptions}
            currencyCode={currencyCode}
            endEditItem={endEditItem}
            isUpdatingItem={isUpdatingItem}
            updateCart={updateItemInCart}
        />
    );
};

EditItem.propTypes = {
    currencyCode: string,
    endEditItem: func,
    isUpdatingItem: bool,
    item: object,
    updateItemInCart: func
};

export default EditItem;
