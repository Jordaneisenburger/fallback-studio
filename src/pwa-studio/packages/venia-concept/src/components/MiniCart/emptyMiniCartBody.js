import React from 'react';
import { func, string, shape } from 'prop-types';

import { mergeClasses } from '../../classify';
import Trigger from '../Trigger';

import defaultClasses from './emptyMiniCartBody.css';

const EmptyMiniCart = props => {
    const { closeDrawer } = props;

    const classes = mergeClasses(defaultClasses, props.classes);

    return (
        <div className={classes.root}>
            <h3 className={classes.emptyTitle}>
                There are no items in your shopping cart
            </h3>
            <Trigger action={closeDrawer}>
                <span className={classes.continue}>Continue Shopping</span>
            </Trigger>
        </div>
    );
};

EmptyMiniCart.propTypes = {
    classes: shape({
        root: string,
        emptyTitle: string,
        continue: string
    }),
    closeDrawer: func
};

export default EmptyMiniCart;
