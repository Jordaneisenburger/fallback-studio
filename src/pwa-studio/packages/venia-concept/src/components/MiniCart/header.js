import React from 'react';
import { bool, func, shape, string } from 'prop-types';
import { X as CloseIcon } from 'react-feather';

import { mergeClasses } from '../../classify';
import Icon from '../Icon';
import Trigger from '../Trigger';

import defaultClasses from './header.css';

const Header = props => {
    const { closeDrawer, isEditingItem } = props;

    const classes = mergeClasses(defaultClasses, props.classes);
    const title = isEditingItem ? 'Edit Cart Item' : 'Shopping Cart';

    return (
        <div className={classes.root}>
            <h2 className={classes.title}>{title}</h2>
            <Trigger action={closeDrawer}>
                <Icon src={CloseIcon} />
            </Trigger>
        </div>
    );
};

Header.propTypes = {
    classes: shape({
        root: string,
        title: string
    }),
    closeDrawer: func,
    isEditingItem: bool
};

export default Header;
