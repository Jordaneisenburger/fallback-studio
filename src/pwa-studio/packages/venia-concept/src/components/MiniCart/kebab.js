import React, { useCallback, useRef, useState } from 'react';
import { node, shape, string } from 'prop-types';
import { MoreVertical as MoreVerticalIcon } from 'react-feather';

import { useEventListener } from '@magento/peregrine';

import { mergeClasses } from '../../classify';
import Icon from '../Icon';

import defaultClasses from './kebab.css';

const Kebab = props => {
    const { children } = props;

    const kebabRef = useRef(null);
    const [isOpen, setIsOpen] = useState(false);

    const classes = mergeClasses(defaultClasses, props.classes);
    const toggleClass = isOpen ? classes.dropdown_active : classes.dropdown;

    const handleKebabClick = useCallback(() => {
        setIsOpen(!isOpen);
    }, [isOpen]);
    const handleOutsideKebabClick = useCallback(event => {
        // Ensure we're truly outside of the kebab.
        if (!kebabRef.current.contains(event.target)) {
            setIsOpen(false);
        }
    }, []);

    useEventListener(document, 'mousedown', handleOutsideKebabClick);
    useEventListener(document, 'touchend', handleOutsideKebabClick);

    return (
        <div className={classes.root}>
            <button
                className={classes.kebab}
                onClick={handleKebabClick}
                ref={kebabRef}
            >
                <Icon src={MoreVerticalIcon} />
            </button>
            <ul className={toggleClass}>{children}</ul>
        </div>
    );
};

Kebab.propTypes = {
    children: node,
    classes: shape({
        dropdown: string,
        dropdown_active: string,
        kebab: string,
        root: string
    })
};

export default Kebab;
