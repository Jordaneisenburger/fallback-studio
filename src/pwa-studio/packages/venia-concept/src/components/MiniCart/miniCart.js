import React from 'react';
import { arrayOf, bool, func, number, object, shape, string } from 'prop-types';

import Body from './body';
import Footer from './footer';
import Header from './header';
import Mask from './mask';
import defaultClasses from './miniCart.css';

import { mergeClasses } from '../../classify';
import getCurrencyCode from '../../util/getCurrencyCode';

const MiniCart = props => {
    // Props.
    const {
        beginEditItem,
        cancelCheckout,
        cart,
        closeDrawer,
        endEditItem,
        isCartEmpty,
        isMiniCartMaskOpen,
        isOpen,
        removeItemFromCart,
        updateItemInCart
    } = props;
    const { editItem, isEditingItem, isLoading, isUpdatingItem } = cart;

    // Members.
    const classes = mergeClasses(defaultClasses, props.classes);
    const currencyCode = getCurrencyCode(cart);
    const cartItems = cart.details.items;
    const numItems = cart.details.items_qty;
    const rootClass = isOpen ? classes.root_open : classes.root;
    const subtotal = cart.totals.subtotal;

    const showFooter = !(isCartEmpty || isLoading || isEditingItem);
    const footer = showFooter ? (
        <Footer
            cart={cart}
            currencyCode={currencyCode}
            isMiniCartMaskOpen={isMiniCartMaskOpen}
            numItems={numItems}
            subtotal={subtotal}
        />
    ) : null;

    return (
        <aside className={rootClass}>
            <Header closeDrawer={closeDrawer} isEditingItem={isEditingItem} />
            <Body
                beginEditItem={beginEditItem}
                cartItems={cartItems}
                closeDrawer={closeDrawer}
                currencyCode={currencyCode}
                editItem={editItem}
                endEditItem={endEditItem}
                isCartEmpty={isCartEmpty}
                isEditingItem={isEditingItem}
                isLoading={isLoading}
                isUpdatingItem={isUpdatingItem}
                removeItemFromCart={removeItemFromCart}
                updateItemInCart={updateItemInCart}
            />
            <Mask isActive={isMiniCartMaskOpen} dismiss={cancelCheckout} />
            {footer}
        </aside>
    );
};

MiniCart.propTypes = {
    beginEditItem: func.isRequired,
    cancelCheckout: func,
    cart: shape({
        details: shape({
            currency: shape({
                quote_currency_code: string
            }),
            items: arrayOf(
                shape({
                    item_id: number,
                    name: string,
                    price: number,
                    product_type: string,
                    qty: number,
                    quote_id: string,
                    sku: string
                })
            ),
            items_qty: number
        }).isRequired,
        editItem: object,
        isEditingItem: bool,
        isLoading: bool,
        isUpdatingItem: bool,
        totals: shape({
            subtotal: number
        }).isRequired
    }).isRequired,
    classes: shape({
        header: string,
        root: string,
        root_open: string,
        title: string
    }),
    closeDrawer: func,
    endEditItem: func.isRequired,
    isCartEmpty: bool,
    isMiniCartMaskOpen: bool,
    isOpen: bool,
    removeItemFromCart: func,
    updateItemInCart: func
};

export default MiniCart;
