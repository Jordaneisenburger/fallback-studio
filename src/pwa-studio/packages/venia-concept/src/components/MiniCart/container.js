import { connect } from '@magento/venia-drivers';
import { isEmptyCartVisible, isMiniCartMaskOpen } from '../../selectors/cart';
import { closeDrawer } from '../../actions/app';
import {
    beginEditItem,
    endEditItem,
    updateItemInCart,
    removeItemFromCart
} from '../../actions/cart';
import { cancelCheckout } from '../../actions/checkout';

import MiniCart from './miniCart';

const mapStateToProps = state => {
    const { cart } = state;

    return {
        cart,
        isCartEmpty: isEmptyCartVisible(state),
        isMiniCartMaskOpen: isMiniCartMaskOpen(state)
    };
};

const mapDispatchToProps = {
    beginEditItem,
    cancelCheckout,
    closeDrawer,
    endEditItem,
    removeItemFromCart,
    updateItemInCart
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(MiniCart);
