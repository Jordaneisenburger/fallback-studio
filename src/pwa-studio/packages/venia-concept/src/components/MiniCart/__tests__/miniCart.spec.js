import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import MiniCart from '../miniCart';

const renderer = new ShallowRenderer();

const baseProps = {
    beginEditItem: jest.fn(),
    cancelCheckout: jest.fn(),
    cart: {
        details: {
            currency: {
                quote_currency_code: 'NZD'
            },
            items: [
                {
                    item_id: 1,
                    name: 'Unit Test Item',
                    price: 99,
                    product_type: 'configurable',
                    qty: 1,
                    quote_id: '1234',
                    sku: 'SKU'
                }
            ],
            items_qty: 1
        },
        editItem: null,
        isEditingItem: false,
        isLoading: false,
        isUpdatingItem: false,
        totals: {
            subtotal: 99
        }
    },
    closeDrawer: jest.fn(),
    endEditItem: jest.fn(),
    isCartEmpty: false,
    isMiniCartMaskOpen: false,
    isOpen: true,
    removeItemFromCart: jest.fn(),
    updateItemInCart: jest.fn()
};

test('renders the correct tree', () => {
    const tree = renderer.render(<MiniCart {...baseProps} />);

    expect(tree).toMatchSnapshot();
});

test('doesnt render a footer when cart is empty', () => {
    const props = {
        ...baseProps,
        isCartEmpty: true
    };

    renderer.render(<MiniCart {...props} />);
    const result = renderer.getRenderOutput();

    const footer = result.props.children[3];
    expect(footer).toBeNull();
});

test('doesnt render a footer when cart is editing', () => {
    const props = {
        ...baseProps,
        cart: {
            ...baseProps.cart,
            isEditingItem: true
        }
    };

    renderer.render(<MiniCart {...props} />);
    const result = renderer.getRenderOutput();

    const footer = result.props.children[3];
    expect(footer).toBeNull();
});

test('doesnt render a footer when cart is loading', () => {
    const props = {
        ...baseProps,
        cart: {
            ...baseProps.cart,
            isLoading: true
        }
    };

    renderer.render(<MiniCart {...props} />);
    const result = renderer.getRenderOutput();

    const footer = result.props.children[3];
    expect(footer).toBeNull();
});
