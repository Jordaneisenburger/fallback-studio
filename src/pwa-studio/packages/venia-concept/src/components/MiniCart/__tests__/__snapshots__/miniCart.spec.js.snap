// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the correct tree 1`] = `
<aside>
  <Header
    closeDrawer={[MockFunction]}
    isEditingItem={false}
  />
  <Body
    beginEditItem={[MockFunction]}
    cartItems={
      Array [
        Object {
          "item_id": 1,
          "name": "Unit Test Item",
          "price": 99,
          "product_type": "configurable",
          "qty": 1,
          "quote_id": "1234",
          "sku": "SKU",
        },
      ]
    }
    closeDrawer={[MockFunction]}
    currencyCode="NZD"
    editItem={null}
    endEditItem={[MockFunction]}
    isCartEmpty={false}
    isEditingItem={false}
    isLoading={false}
    isUpdatingItem={false}
    removeItemFromCart={[MockFunction]}
    updateItemInCart={[MockFunction]}
  />
  <MiniCartMask
    dismiss={[MockFunction]}
    isActive={false}
  />
  <Footer
    cart={
      Object {
        "details": Object {
          "currency": Object {
            "quote_currency_code": "NZD",
          },
          "items": Array [
            Object {
              "item_id": 1,
              "name": "Unit Test Item",
              "price": 99,
              "product_type": "configurable",
              "qty": 1,
              "quote_id": "1234",
              "sku": "SKU",
            },
          ],
          "items_qty": 1,
        },
        "editItem": null,
        "isEditingItem": false,
        "isLoading": false,
        "isUpdatingItem": false,
        "totals": Object {
          "subtotal": 99,
        },
      }
    }
    currencyCode="NZD"
    isMiniCartMaskOpen={false}
    numItems={1}
    subtotal={99}
  />
</aside>
`;
